sysvabi64/sysvabi64.rst

```diff
@@ -0,0 +1,1076 @@
+    adr     x0, lbdst     // R_AARCH64_ADR_PREL_LO21 lbdst
+    adr     x1, lbsrc     // R_AARCH64_ADR_PREL_LO21 lbsrc
+    ldrb    w1, [x1]
+    strb    w1, [x0]
+    adr     x1, .LANCHOR0 // R_AARCH64_ADR_PREL_LO21 .LANCHOR0
+    str     x0, [x1, 8]
+    ret
+
+  getweak:
+    ldr     x0, .LC0
+    ret
+    .align 3
+  .LC0:
+    .xword  weakref       // R_AARCH64_ABS64 weakref
+
+  .bss
+  .LANCHOR0
+  ldst:
+        .zero   1
+        .zero   7
+  lbdst:
+        .zero   65536
+
+
+.. note::
+
+   There is no obvious advantage of ``adr/ldr`` over the
+   ``adrp/ldr[#:lo12]`` used by the small code-model.
+
+   Benefits over the small model are the ability to use the
+   PC-relative loads of 32 and 64-bit data and a single ADR
+   instruction to compute the address of any code or data symbol.
+
+   An external symbol reference with a weak binding must always be
+   output using the large absolute addressing model because if the
+   program is linked to run at a virtual address higher than 1MiB,
+   then it will not be possible to generate 0 as the address of an
+   undefined weak symbol using the ADR instruction.
+
+Small Code Model
+^^^^^^^^^^^^^^^^
+
+The small memory model uses the 2 instruction page + offset PC relative
+addressing to access data which is all within +/- 4GiB of the code.
+
+.. code-block:: asm
+
+  foo:
+    adrp    x0, dst                      // R_AARCH64_ADR_PREL_PG_HI21   dst
+    add     x1, x0, :lo12:dst            // R_AARCH64_LDST64_ABS_LO12_NC dst
+    adrp    x2, src                      // R_AARCH64_ADR_PREL_PG_HI21   src
+    ldrb    w2, [x2, #:lo12:src]         // R_AARCH64_LDST64_ABS_LO12_NC src
+    strb    w2, [x0, #:lo12:dst]         // R_AARCH64_LDST64_ABS_LO12_NC src
+    adrp    x0, ptr                      // R_AARCH64_ADR_PREL_PG_HI21   ptr
+    str     x1, [x0, #:lo12:ptr]         // R_AARCH64_LDST64_ABS_LO12_NC ptr
+    ret
+
+  bar:
+    adrp    x1, .LANCHOR0                // R_AARCH64_ADR_PREL_PG_HI21   .LANCHOR0
+    add     x0, x1, :lo12:.LANCHOR0      // R_AARCH64_LDST64_ABS_LO12_NC .LANCHOR0
+    ldrb    w2, [x0, 1]
+    strb    w2, [x1, #:lo12:.LANCHOR0]   // R_AARCH64_LDST64_ABS_LO12_NC .LANCHOR0
+    str     x0, [x0, 8]
+    ret
+
+  baz:
+    adrp    x0, lbdst                    // R_AARCH64_ADR_PREL_PG_HI21   lbdst
+    add     x1, x0, :lo12:lbdst          // R_AARCH64_LDST64_ABS_LO12_NC lbdst
+    adrp    x2, lbsrc                    // R_AARCH64_ADR_PREL_PG_HI21   lbsrc
+    ldrb    w2, [x2, #:lo12:lbsrc]       // R_AARCH64_LDST64_ABS_LO12_NC lbsrc
+    strb    w2, [x0, #:lo12:lbdst]       // R_AARCH64_LDST64_ABS_LO12_NC lbdst
+    adrp    x0, .LANCHOR0+8              // R_AARCH64_ADR_PREL_PG_HI21   .LANCHOR0 + 8
+    str     x1, [x0, #:lo12:.LANCHOR0+8] // R_AARCH64_LDST64_ABS_LO12_NC .LANCHOR0 + 8
+    ret
+
+  getweakaddr:
+    adrp    x0, .LC0                     // R_AARCH64_ADR_PREL_PG_HI21   .LC0
+    ldr     x0, [x0, #:lo12:.LC0]        // R_AARCH64_LDST64_ABS_LO12_NC .LC0
```

Comment:
I'm a bit reluctant to do that for the first commit. Using real world compiler output makes it less likely that mistakes will keep in. Alternatives here are:
1.) clang uses PIC addressing (a GOT relative reference). Still a literal load but at least it would be one per program and not in the instruction cache.
2.) Use large model addressing via MOV instructions.

I'll keep it as it is for now. We can alter with a follow up patch.

Reply:
It doesn't look like this is real world compiler output - GCC places the LC0 constant in rodata which isn't shown here.

Reply:
I'll update, I missed the .section in the disassembly.